aws/middleware/user_agent.go

```diff
@@ -0,0 +1,116 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/awslabs/smithy-go/httpbinding"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+)
+
+// UserAgent is an interface for modifying the User-Agent of a request.
+type UserAgent interface {
+	AddKey(name string)
+	AddKeyValue(name, version string)
+}
+
+// RequestUserAgent is a build middleware that set the User-Agent for the request.
+type RequestUserAgent struct {
+	uab *httpbinding.UserAgentBuilder
+}
+
+// NewRequestUserAgent returns a new RequestUserAgent which will set the User-Agent for the request.
+//
+// Default Example:
+//   aws-sdk-go/2.3.4 GOOS/linux GOARCH/amd64 GO/go1.14
+func NewRequestUserAgent() *RequestUserAgent {
+	uab := httpbinding.NewUserAgentBuilder()
+	uab.AddKeyValue(aws.SDKName, aws.SDKVersion)
+	uab.AddKeyValue("GOOS", runtime.GOOS)
+	uab.AddKeyValue("GOARCH", runtime.GOARCH)
+	uab.AddKeyValue("GO", runtime.Version())
+	return &RequestUserAgent{uab: uab}
+}
+
+// AddUserAgentKey retrieves the RequestUserAgent from the provided stack, or initializes
+func AddUserAgentKey(callback func(interface {
+	AddKey(string)
+}) error) func(*middleware.Stack) error {
+	return func(stack *middleware.Stack) error {
+		requestUserAgent, err := getOrAddRequestUserAgent(stack)
+		if err != nil {
+			return err
+		}
+		return callback(requestUserAgent)
+	}
+}
```

Comment:
I think the callback pattern is equivalent to invoking `requestUserAgent.AddKey(v)` if the `AddUserAgentKey` function was updated to take `v string`.

but is potentially a little bit similiar because I wouldn't need to define another function closure. (This also changes `AddKey` to return an error.

```go
func AddUserAgentKey(key string) func(*middleware.Stack) error {
     return func(stack *middleware.Stack) error {
          requestUserAgent, err := getOrAddRequestUserAgent(stack)
          if err != nil { return err }

          return requestUserAgent.AddKey(key)
}
```

with the usage of the two patterns being:

```go
client.DoThing(ctx, params, middleware.AddUserAgentKey(func(ua interface{AddKey(string)}) error {
       ua.AddKey("foo abc")
       return nil
}))
```

Vs without the callback

```go
client.DoThing(ctx, params, middleware.AddUserAgentKey("foo abc"))
```